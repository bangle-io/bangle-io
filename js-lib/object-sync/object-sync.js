export function objectSync(obj, emitChange) {
  const allowedKeys = new Set(Object.keys(obj));

  obj = Object.assign({}, obj);
  const valueCheck = (value) => {
    if (
      typeof value === 'string' ||
      typeof value === 'number' ||
      typeof value === 'boolean' ||
      value == null
    ) {
      return true;
    }
    console.log(value);
    throw new Error('Invalid value type ' + typeof value);
  };

  Object.values(obj).forEach((val) => {
    valueCheck(val);
  });

  let counter = 0;

  let changeListeners = [];

  const updateProp = (target, key, value) => {
    valueCheck(value);
    if (!allowedKeys.has(key)) {
      throw new Error('Invalid key ' + key);
    }

    const existingValue = Reflect.get(target, key);
    if (existingValue === value) {
      return true;
    }
    counter++;
    Reflect.set(target, key, value);
    notifyListeners();
    return true;
  };

  const proxy = new Proxy(obj, {
    set(target, propKey, value, receiver) {
      updateProp(target, propKey, value);
      emitChange({
        type: 'UPDATE',
        payload: {
          key: propKey,
          value: Reflect.get(target, propKey),
        },
      });
      return true;
    },
    deleteProperty() {
      throw new Error('Not allowed');
    },
  });

  let notifyListeners = () => {
    changeListeners.forEach((fn) => fn(proxy));
  };

  return {
    // a number that can be used to check if the object has updated since last time
    get counter() {
      return counter;
    },
    // callback that updates the current object to match
    // with the one you are syncing with.
    applyForeignChange: ({ type, payload: { key, value } }) => {
      if (type === 'UPDATE') {
        updateProp(obj, key, value);
      } else {
        throw new Error('Unknown type of change');
      }
    },
    proxy,
    register: (fn) => {
      if (changeListeners.includes(fn)) {
        return;
      }
      changeListeners.push(fn);
    },
    deregister: (fn) => {
      changeListeners = changeListeners.filter((f) => f !== fn);
    },
  };
}
