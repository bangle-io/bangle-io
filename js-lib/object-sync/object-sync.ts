export type ObjectSyncEventType<T> = {
  type: 'UPDATE';
  payload: {
    key: string | Symbol;
    value: T;
  };
};

export type ObjectSyncCallback<T> = (param: {
  counter: number;
  appStateValue: { [key: string]: T };
}) => void;

export function objectSync<T>(
  obj: { [key: string]: T },
  { emitChange = (p: ObjectSyncEventType<T>) => {} } = {},
) {
  const allowedKeys = new Set(Object.keys(obj));

  obj = Object.assign({}, obj);
  const valueCheck = (value) => {
    if (
      typeof value === 'string' ||
      typeof value === 'number' ||
      typeof value === 'boolean' ||
      value == null
    ) {
      return true;
    }
    throw new Error('Invalid value type ' + typeof value);
  };

  Object.values(obj).forEach((val) => {
    valueCheck(val);
  });

  let counter = 0;

  let changeListeners: Array<ObjectSyncCallback<T>> = [];

  const updateProp = (target, key: string | Symbol, value: T) => {
    if (typeof key !== 'string') {
      throw new Error('Key much be string type');
    }

    valueCheck(value);

    if (!allowedKeys.has(key)) {
      throw new Error('Invalid key ' + key);
    }

    const existingValue: T = Reflect.get(target, key);
    if (existingValue === value) {
      return true;
    }
    counter++;
    Reflect.set(target, key, value);
    notifyListeners();
    return true;
  };

  const proxy = new Proxy(obj, {
    set(target, propKey, value, receiver) {
      updateProp(target, propKey, value);
      emitChange({
        type: 'UPDATE',
        payload: {
          key: propKey,
          value: Reflect.get(target, propKey),
        },
      });
      return true;
    },
    deleteProperty() {
      throw new Error('Not allowed');
    },
  });

  let notifyListeners = () => {
    changeListeners.forEach((fn: ObjectSyncCallback<T>) =>
      fn({ counter, appStateValue: proxy }),
    );
  };

  return {
    // a number that can be used to check if the object has updated since last time
    get counter() {
      return counter;
    },
    // callback that updates the current object to match
    // with the one you are syncing with.
    applyForeignChange: ({
      type,
      payload: { key, value },
    }: ObjectSyncEventType<T>) => {
      if (type === 'UPDATE') {
        updateProp(obj, key, value);
      } else {
        throw new Error('Unknown type of change');
      }
    },
    appStateValue: proxy,
    registerListener: (fn: ObjectSyncCallback<T>) => {
      if (changeListeners.includes(fn)) {
        return;
      }
      changeListeners.push(fn);
    },
    deregisterListener: (fn: ObjectSyncCallback<T>) => {
      changeListeners = changeListeners.filter((f) => f !== fn);
    },
  };
}
