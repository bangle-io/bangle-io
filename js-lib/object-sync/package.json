{
  "name": "@bangle.io/object-sync",
  "version": "0.0.0",
  "authors": [
    {
      "name": "Kushan Joshi",
      "email": "0o3ko0@gmail.com",
      "web": "http://github.com/kepta"
    }
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/bangle-io/bangle-io.git",
    "directory": "js-lib/object-sync"
  },
  "bugs": {
    "url": "https://github.com/bangle-io/bangle-io/issues"
  },
  "main": "index.ts",
  "module": "index.ts",
  "publishConfig": {},
  "dependencies": {
    "tslib": "^2.3.1"
  }
}
