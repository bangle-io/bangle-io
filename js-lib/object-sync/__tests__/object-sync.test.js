import { objectSync } from '../object-sync';

function sleep(t = 10) {
  return new Promise((res) => setTimeout(res, t));
}
test('syncing works', async () => {
  let objA = { test: undefined };
  let objB = { test: undefined };

  let a = objectSync(objA, {
    emitChange: (p) => {
      return sleep(5).then(() => {
        b.applyForeignChange(p);
      });
    },
  });
  let b = objectSync(objB, {
    emitChange: (p) => {
      return sleep(5).then(() => {
        a.applyForeignChange(p);
      });
    },
  });

  a.appStateValue.test = 123;
  await sleep(20);
  expect(b.appStateValue.test).toBe(123);
  expect(a.counter).toBe(1);
  b.appStateValue.test = 124;
  await sleep(20);

  expect(b.counter).toBe(2);
  expect(a.counter).toBe(2);

  expect(a.appStateValue.test).toBe(124);
});

test('throws error if not primitive type', async () => {
  let objA = {};
  let objB = {};

  let a = objectSync(objA, {
    emitChange: (p) => {
      return sleep(5).then(() => {
        b.applyForeignChange(p);
      });
    },
  });
  let b = objectSync(objB, {
    emitChange: (p) => {
      return sleep(5).then(() => {
        a.applyForeignChange(p);
      });
    },
  });

  expect(() => {
    a.appStateValue.test = {};
  }).toThrowErrorMatchingSnapshot();
});

test('throws error if initial not primitive type', async () => {
  let objA = { b: 12 };
  let objB = {};

  let a = objectSync(objA, {
    emitChange: (p) => {
      return sleep(5).then(() => {
        b.applyForeignChange(p);
      });
    },
  });
  let b = objectSync(objB, {
    emitChange: (p) => {
      return sleep(5).then(() => {
        a.applyForeignChange(p);
      });
    },
  });

  expect(() => {
    a.appStateValue.test = {};
  }).toThrowErrorMatchingSnapshot();
});

test('throws error if unknown key', async () => {
  let objA = { b: 12 };

  let a = objectSync(objA);
  // this should be fine
  a.appStateValue.b = 13;

  expect(() => {
    a.appStateValue.c = 13;
  }).toThrowErrorMatchingSnapshot();
});
