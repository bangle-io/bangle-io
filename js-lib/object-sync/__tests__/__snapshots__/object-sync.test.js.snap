// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws error if initial not primitive type 1`] = `"Invalid value type object"`;

exports[`throws error if not primitive type 1`] = `"Invalid value type object"`;

exports[`throws error if unknown key 1`] = `"Invalid key c"`;
